programs/config_api/src/config_state.rs

```diff
@@ -0,0 +1,203 @@
+use solana_sdk::account::KeyedAccount;
+use solana_sdk::native_program::ProgramError;
+
+#[derive(Debug, Default, Serialize, Deserialize, PartialEq, Clone)]
+pub struct ConfigState {
+    // Sequence number that increments on every update, will never decrement
+    pub seq: u64,
+
+    // Configuration data
+    pub data: Vec<u8>,
+}
+
+impl ConfigState {
+    pub fn new(data_len: u64) -> Self {
+        Self {
+            seq: 0,
+            data: vec![0; data_len as usize],
+        }
+    }
+
+    pub fn max_size(data_len: u64) -> u64 {
+        serialized_size(&ConfigState::new(data_len)).unwrap()
+    }
+
+    pub fn deserialize(input: &[u8]) -> Result<Self, ProgramError> {
+        deserialize(input).map_err(|_| ProgramError::InvalidAccountData)
+    }
+
+    pub fn serialize(&self, output: &mut [u8]) -> Result<(), ProgramError> {
+        serialize_into(output, self).map_err(|err| match *err {
+            ErrorKind::SizeLimit => ProgramError::InvalidAccountData,
+            _ => ProgramError::GenericError,
+        })
+    }
+
+    pub fn store(&mut self, new_config_state: ConfigState) -> Result<(), ProgramError> {
+        if new_config_state.seq <= self.seq {
+            warn!(
+                "sequence number old ({} <= {})",
+                new_config_state.seq, self.seq
+            );
+            Err(ProgramError::InvalidArgument)?;
+        }
+
+        *self = new_config_state;
+        Ok(())
+    }
+}
+
+pub fn process(
+    keyed_accounts: &mut [KeyedAccount],
+    new_config_state: ConfigState,
+) -> Result<(), ProgramError> {
+    if !check_id(&keyed_accounts[0].account.owner) {
+        error!("account[0] is not assigned to the config program");
+        Err(ProgramError::InvalidArgument)?;
+    }
+
+    if keyed_accounts[0].signer_key().is_none() {
+        error!("account[0] should sign the transaction");
+        Err(ProgramError::InvalidArgument)?;
+    }
+    let config_state = ConfigState::deserialize(&keyed_accounts[0].account.data);
+    if let Ok(mut config_state) = config_state {
+        config_state.store(new_config_state)?;
+        config_state.serialize(&mut keyed_accounts[0].account.data)?;
+    } else {
+        error!("account[0] does not contain valid account data");
+        Err(ProgramError::InvalidAccountData)?;
+    }
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::id;
+    use solana_sdk::account::Account;
+    use solana_sdk::pubkey::Pubkey;
+    use solana_sdk::signature::{Keypair, KeypairUtil};
+
+    fn create_config_account(data_len: u64, lamports: u64) -> Account {
+        let space = ConfigState::max_size(data_len) as usize;
+        Account::new(lamports, space, &id())
+    }
+
+    #[test]
+    fn test_process_store_ok() {
+        solana_logger::setup();
+        let config_account_pubkey = Keypair::new().pubkey();
+        let mut config_account = create_config_account(42, 1);
+        let mut keyed_accounts = [KeyedAccount::new(
```

Comment:
While these are pure unit tests, I've found the KeyedAccounts quite tedious to work with and not all that great in weeding out the most common bugs, which for me, has been accidentally breaking the runtime's rules:

Check out the tests in: programs/budget/src/budget_program.rs

Reply:
Cool yeah that’s totally nicer